src/qutip_qip/decompose/decompose_two_qubit_gate.py

```diff
@@ -0,0 +1,298 @@
+        gate_list[4],
+        Pauli_X,
+        CNOT_ctrl_1,
+        Pauli_X,
+        gate_list[6],
+        phase_gate,
+    )
+
+
+def _decompose_to_two_level_arrays(input_gate):
+    """Output of two qubit gate in terms of two-level numpy arrays.
+    """
+    check_gate(input_gate, num_qubits=2)
+    input_array = input_gate.full()
+
+    # Calculate the two level numpy arrays
+    array_list = []
+    index_list = [[0, 1], [0, 2], [0, 3], [1, 2], [1, 3]]
+    for i in range(len(index_list)):
+        index_1, index_2 = index_list[i]
+
+        # Values of single qubit U forming the two level unitary
+        a = input_array[index_1][index_1]
+        a_star = np.conj(a)
+        b = input_array[index_2][index_1]
+        b_star = np.conj(b)
+        norm_constant = cmath.sqrt(np.absolute(a*a_star)+np.absolute(b*b_star))
+
+        # Create identity array and then replace with above values for index_1
+        # and index_2
+        U_two_level = np.identity(4, dtype=complex)
+        U_two_level[index_1][index_1] = a_star/norm_constant
+        U_two_level[index_2][index_1] = b/norm_constant
+        U_two_level[index_1][index_2] = b_star/norm_constant
+        U_two_level[index_2][index_2] = -a/norm_constant
+
+        # Change input by multiplying by above two-level
+        input_array = np.dot(U_two_level, input_array)
+
+        # U dagger to calculate the gates
+        U__two_level_dagger = np.transpose(np.conjugate(U_two_level))
+        U__two_level_dagger = Qobj(U__two_level_dagger, dims=[[2, 2], [2, 2]])
+        array_list.append(U__two_level_dagger)
+
+    # for U6 - multiply input array by U5 and take dagger
+    U6_dagger = input_array
+    array_list.append(Qobj(U6_dagger, dims=[[2, 2], [2, 2]]))
+    return(array_list)
+
+
+def decompose_two_qubit_to_two_level_unitary(input_gate):
+    """Decomposes input two-qubit unitary into two-level unitary gate objects.
+
+    Returns
+    ----------
+    tuple: (two_level_dictionary, two_level_gates)
+
+    two_level_dictionary
+        Dictionary containing string labels to insert user-defined gates in
+        a quantum circuit.
+
+    two_level_gates
+        Gates describing the decomposition of the two-qubit input gate.
+    """
+    check_gate(input_gate, num_qubits=2)
+    array_list = _decompose_to_two_level_arrays(input_gate)
+    user_gates = {
+        "U1": array_list[0],
+        "U2": array_list[1],
+        "U3": array_list[2],
+        "U4": array_list[3],
+        "U5": array_list[4],
+        "U6": array_list[5],
+        }
+
+    U1_gate = Gate("U1", targets=[0, 1])
+    U2_gate = Gate("U2", targets=[0, 1])
+    U3_gate = Gate("U3", targets=[1, 0])
+    U4_gate = Gate("U4", targets=[0, 1])
+    U5_gate = Gate("U5", targets=[0, 1])
+    U6_gate = Gate("U6", targets=[0, 1])
+
+    gate_list = (U6_gate, U5_gate, U4_gate, U3_gate, U2_gate, U1_gate)
+
+    return(user_gates, gate_list)
+
+
+def decompose_two_qubit_to_CNOT_and_single_qubit_gates(input_gate):
```

Comment:
Output of this function is not what it is expected to be. For this reason, all of the smaller gates for U6,...,U1 have been defined separately. A for loop will be created for these once the output matches exactly to the input. 

Need to recalculate the target and control qubits because they were different for the two level circuit output function. The output of the two level output matches exactly to the input. 

The hope is to check my calcualtions with the targets and controls from gray code function's output.